### DISCUSSION TITLE
Inputs / Outputs

### DISCUSSION THREAD

**Discussion by gnodar01** [bilayer-containers/bilayers#80] [Ideas]
As we start building out inputs and outputs, we'll want to take a look at what others have done.

The [2nd International Workshop on FAIR Containerized Computational Software](https://www.nist.gov/news-events/events/2nd-international-workshop-fair-containerized-computational-software) held by NIST is a great starting point. The impetus of the workshop:

> containerized software tools must be interoperable as they are chained into workflows

and

> The main goal for the workshop is to establish a community consensus on creating interoperable containerized computational tools that can be chained into scientific workflows/pipelines and executed over large image collections regardless of the cloud infrastructure components.  

Also the report from the [1st workshop](https://nvlpubs.nist.gov/nistpubs/ir/2020/NIST.IR.8297.pdf)

Both workshops were focused on Microscopy Image Analyses as the first, and primary, area for exploring the ideas. Hence commonalities with what we're doing.

The primary outcome of the workshop was to

> establish fields in a manifest file accompanying each containerized software tool

The manifest file is a json file analogous to our config.yaml spec.

Day one was focused on Inputs/Outputs (what we're working on). Day two was focused on GUIs (what we have already worked on). Day three was on Hardware Requirements (what we have not really thought through yet).

[This is the repo](https://github.com/usnistgov/fair-chain-compute-container) containing the current state of their "mainfest" (spec).

I think it has a lot of commonalities with what we want (on some things we don't). [This presentation](https://www.nist.gov/system/files/documents/2023/12/08/DayOne_Plugin-manifest_MyleneSimon_0.pdf) is a good summary.

Other things we should look at:

- [Common Workflow Language](https://www.commonwl.org/)
- [Workflow Description Language](https://openwdl.org/)
- [Galaxy inputs and outputs](https://docs.galaxyproject.org/en/latest/dev/schema.html)
- ...

**Comment by rajavishah** [comment]
With all the back-n-forth conversation, currently our `inputs`, and `outputs` are now somewhere blended in with parameters. So now, `type: files` in `parameters`, `display_only` and `results` can be more vividly thought in below 5 categories. Those are `image`, `array`, `measurement`, `file` and `executable`. 
And now, all the `parameters` which required `type: files` can be either placed under `inputs` or `outputs` wherever relevant. Most of the use case which I have faced for current 3 algorithms are just needed to move `type: files` to `inputs` only. No major use case has been found for moving `files` to `outputs`. 

### 1. Speaking more about `inputs`:
`inputs` would definitely follow the same schema as `parameters` but also a bunch of new flags are added. So, here is the final skeleton of `inputs`:
```
- name: input_xyz_type
    type: ("image" | "measurement" | "array" | "file" | "executable")
    label: "Input Images" 
    description: "these are accepted input_images for xyz algorithm"
    cli_tag: "--input_images" # tag to be used in CLI
    cli_order: Integer # default is 0. Could be positive, negative or zero. If negative, then it will be appended at the end of the command. BUT IT IS OPTIONAL TO PROVIDE THIS FLAG.
    default: ("single" | "directory") # if it is a single file or a folder
    optional: (True | False) # if input is optional
    format: # can include all those relevant formats supported by the tool
     - abc
     - xyz
     - ...
    folder_name: "/path/to/input_images" # folder_path
    file_count: ("single" | "multiple") # if it accepts single file or multiple files. If multiple, then it will be a folder
    section_id: "inputs" # section_id to group the inputs together
    mode: ("beginner" | "advanced")
```

Now, if `type == image`. 5 more flags must be added. i.e. `subtype`, `depth`, `timepoints`, `tiled`, `pyramidal`. So, below is a mock example of how it would look like if `type: image`. 

```
- name: input_images
    type: image
    label: "Input Images"
    subtype: # ONLY IFF TYPE==IMAGE. Can list all those relevant subtypes out of these 4, supported by the tool
      - grayscale
      - color
      - binary
      - labeled
    description: "these are accepted input_images for xyz algorithm"
    cli_tag: "--input_images" # tag to be used in CLI
    cli_order: Integer # default is 0. Could be positive, negative or zero. If negative, then it will be appended at the end of the command. BUT IT IS OPTIONAL TO PROVIDE THIS FLAG.
    default: ("single" | "directory") # if it is a single file or a folder
    optional: (True | False) # if input is optional
    format: # can include all those relevant formats supported by the tool
      - tif
      - png
      - jpg
      - jpeg
      - tiff
      - ometiff
    folder_name: "/path/to/input_images" # folder_path
    file_count: ("single" | "multiple") # if it accepts single file or multiple files. If multiple, then it will be a folder
    section_id: "inputs" # section_id to group the inputs together
    mode: ("beginner" | "advanced")
    # Extra flags ONLY IFF TYPE==IMAGE
    depth: (True | False) # ONLY IF TYPE IS IMAGE. If tool accepts images, with z-dimension i.e. depth, then set it to True
    timepoints: (True | False) # ONLY IF TYPE IS IMAGE. If tool accepts images, with t-dimension i.e. timepoints, then set it to True
    tiled: (True | False) # ONLY IF TYPE IS IMAGE. If tool accepts images, with tiled format, then set it to True
    pyramidal: (True | False) # ONLY IF TYPE IS IMAGE. If tool accepts images, with pyramidal format, then set it to True
  
```

In the overall structure some of the things I've tweaked. Hence, dumping the thought process - if we need may circle back to those choices. 
#### a. 
We are in a way providing support for `types: files` from `parameters` and also, `inputs` is useful for integrating spec in any workflow we needed some more `flags`. Hence, fetching from `parameters` if `type: files`; the skeleton would look something like 
```
-   name: 
    type: files
    label: ""
    description: ""
    file_count: ("single" | "multiple")
    default: ("single" | "directory")
    cli_tag: ""
    cli_order: 0
    optional: True
    section_id: ""
    folder_name: ""
    mode: ""
```

Hence, flag `default` and `file_count` could help us decide whether it's `collection (i.e. folder)` or not. Therefore, removed `collection` flag from `inputs`, `outputs` schema.
Previously, we had `path` in `inputs`, `outputs`. Now, I suppose `folder_name` from `parameters` help us gain the same information. So, removed that too.

#### b. 
if `type: image` we decided on including `z`, `t` and `subtype`, `tiled`, `pyramidal`. But LinkML linter was throwing a warning `standard_naming` if we use `z` and `t`. Hence, renamed those to `z --> depth`, `t --> timepoints`.

### 2. Removing `types: files` support from logic of `Interface`
Ideally, I shouldn't have support of `type: files`, but it's still present at both `LinkML` side and in core logic of `src/build/parse/gradio_template.py.j2` and `src/build/parse/jupyter_template.py.j2`. 
- as of now, only `results` uses it to generate the UI for `gradio`. Definitely, that could be replaced with `files --> file`. And the easiest way to unsupport `type: files` is to remove from linkML schema file, hence, it flags the `spec` file before it can proceed further. The only reason why I have still kept the logic is I haven't thought any test_cases where `types: files` could provide additional value and may be needed. 

### 3. `outputs` are currently not rendered for generating UI or `cli_command`
The current algorithms don't have any specific use case to fetch `components` from `outputs` and translate it to UI components. I feel the use of `outputs` would be only significant in some workflow. If we parse `outputs` then there would be a potential conflict on UI - 2 components for same `flag`. For eg. `--save_png` is actually `type=checkbox` but under `outputs` we would specify as `type=image`.

More detailed version of how an ideal `spec file` should be is at: `src/build/algorithms/example_algorithm/config.yaml`. (Maybe worth checking `updated_ip_op` branch if it's not merged to `main`)

**Comment by rajavishah** [comment]
> 2. Removing types: files support from logic of Interface
Ideally, I shouldn't have support of type: files, but it's still present at both LinkML side and in core logic of src/build/parse/gradio_template.py.j2 and src/build/parse/jupyter_template.py.j2.
>
> as of now, only results uses it to generate the UI for gradio. Definitely, that could be replaced with files --> file. And the easiest way to unsupport type: files is to remove from linkML schema file, hence, it flags the spec file before it can proceed further. The only reason why I have still kept the logic is I haven't thought any test_cases where types: files could provide additional value and may be needed.

This has changed! Now, I've completely removed support of type==files from `parameters` and `display_only`. In future, if we wish to have `type==files` support or for that matter, wish to add any new UI type support - make sure to reflect changes in all the following files
1. validate_schema.yaml
2. gradio_template.py.j2
3. jupyter_template.py.j2
4. Also, refine the test cases under tests folder
5. Maybe(?) tweak/verify all the existing algorithm's `config.yaml` file - whether anything is breaking!

And, since, `results` section is removed (which was helpful in generating the `Download outputs` component in the Gradio app), the section is directly generated in gradio_template.py.j2 template. (See, #86)
